### PR TITLE
Fixes #4040, #5378 - signal has no `connect` member

### PR DESCRIPTION
**Pull request by adam-grant-hendry**
## Steps

- [x] For new features or bug fixes, add a ChangeLog entry describing what your PR does.
- [x] Write a good description on what the PR does.

## Description

PR #900 contained [a comment](https://github.com/PyCQA/astroid/pull/900#issuecomment-804492415) with a suggested fix that solved the problem experienced by users.

To reiterate, for PySide2 5.15.2+ and PySide6, `pylint` complains that signals have no `connect` member, even though all objects that inherit from `QObject` have a `connect` signal. Issue #900 explains PySide2 version 5.15.2 changed the `Signal` class to implement the descriptor protocol, hence `pylint` began detecting signals as `FunctionDef`s instead of `ClassDef`s causing them not to be detected and patched anymore. The issue persisted in PySide6, hence [the comment that solved the problem](https://github.com/PyCQA/astroid/pull/900#issuecomment-804492415) was made.

Unfortunately, the reviewer and original submitter are no longer active with the project, so it was requested a new PR be made referencing the old one that contains the fix.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Related Issue

See original PR #900 

Closes https://github.com/PyCQA/pylint/issues/4040 and https://github.com/PyCQA/pylint/issues/5378

### CODE REVIEW

File: ChangeLog [hunk 1]
```diff
@@ -3283,9 +3166,6 @@ Release date: 2006-03-10
 
     * fix a possible NameError in Instance.infer_call_result
 
-
-
-
```
Comment (jacobtylerwalls) [inline ChangeLog:3288]: Please only add your changelog to this file.

**Review by jacobtylerwalls — CHANGES_REQUESTED** [review]
Thanks for resurrecting this stalled issue! Would you be able to prepare a version of this PR that avoids unrelated formatting changes?

File: .gitignore [hunk 2]
```diff
@@ -15,3 +15,6 @@ astroid.egg-info/
 .pytest_cache/
 .mypy_cache/
 venv
+.venv/
```
Comment (Pierre-Sassoulas) [inline .gitignore:18]:
```suggestion
```
This is specific to your dev env and should be in a global gitignore on your side. (You won't have to add this in all your project)

File: .gitignore [hunk 3]
```diff
@@ -15,3 +15,6 @@ astroid.egg-info/
...
+.vscode/
```
Comment (Pierre-Sassoulas) [inline .gitignore:19]:
```suggestion

File: script/.contributors_aliases.json [hunk 4]
```diff
@@ -169,5 +169,9 @@
   "ville.skytta@iki.fi": {
     "mails": ["ville.skytta@iki.fi", "ville.skytta@upcloud.com"],
     "name": "Ville Skyttä"
+  },
+  "adam.grant.hendry@gmail.com": {
+    "mails": ["adam.grant.hendry@gmail.com"],
+    "name": "Adam Hendry"
```
Comment (Pierre-Sassoulas) [inline script/.contributors_aliases.json:175]: No need to open another merge request for this imo.

File: setup.cfg [hunk 5]
```diff
@@ -68,7 +68,10 @@ scripts_are_modules = True
 no_implicit_optional = True
 warn_redundant_casts = True
 show_error_codes = True
-enable_error_code = ignore-without-code
+# See python/mypy Issue #10709
```
Comment (Pierre-Sassoulas) [inline setup.cfg:71]: I think we're not ready to activate mypy yet (it's in precommit but everything is deactivated), it's better if we don't mix it with another change.

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 2 above)
Comment (adam-grant-hendry) [reply .gitignore:18]: Ah apologies, yes. I've removed this and put it in my global `.gitignore`.

File: .gitignore (hunk 3 above)
Comment (adam-grant-hendry) [reply .gitignore:19]: Same for above

File: script/.contributors_aliases.json (hunk 4 above)
Comment (adam-grant-hendry) [reply script/.contributors_aliases.json:175]: Let me know if you want me to take it out. I'll leave it in for now. I'm fine either way.

File: .gitignore [hunk 6]
```diff
@@ -15,3 +15,4 @@ astroid.egg-info/
 .pytest_cache/
 .mypy_cache/
 venv
+doc/_build/
```
Comment (adam-grant-hendry) [inline .gitignore:18]: Note that I've left this one in. Let me know if you want me to remove it.

**Comment by adam-grant-hendry** [comment]
> Thanks for resurrecting this stalled issue! Would you be able to prepare a version of this PR that avoids unrelated formatting changes?

Yes, apologies for that. That should be fixed in my last commit.

**Review by Pierre-Sassoulas — COMMENTED** [review]
Thank you for the cleanup, and thank you for ignoring the generated doc, we'tr not giving enough love to the doc in astroid and it shows 😅 Do you think there could be a way to test that without installing dependencies before tests ?

File: astroid/brain/brain_qt.py [hunk 7]
```diff
@@ -71,6 +71,17 @@ def emit(self, signal): pass
     )
 
 
+def _looks_like_pyside2_or_6_signal(node):
```
Comment (DanielNoord) [inline astroid/brain/brain_qt.py:74]:
```suggestion
def _looks_like_pyside2_or_6_signal(node: nodes.FunctionDef) -> bool:
```

**Review by DanielNoord — CHANGES_REQUESTED** [review]

**Comment by adam-grant-hendry** [comment]
> Thank you for the cleanup, and thank you for ignoring the generated doc, we'tr not giving enough love to the doc in astroid and it shows 😅 Do you think there could be a way to test that without installing dependencies before tests ?

Do you mean test that the generated docs in `doc/_build` don't get tracked by `git`?

**Comment by Pierre-Sassoulas** [comment]
I was thinking about test for the qt brain.

File: astroid/brain/brain_qt.py [hunk 8]
```diff
@@ -71,6 +71,17 @@ def emit(self, signal): pass
     )
 
 
+def _looks_like_pyside2_or_6_signal(node: nodes.FunctionDef) -> bool:
+    """Detect a PySide2 or PySide6 signal.  These changed locations as of QT 5ish apparently."""
```
Comment (DanielNoord) [inline astroid/brain/brain_qt.py]:
```suggestion
    """Detect a PySide2 or PySide6 signal.
    
    These changed locations as of QT 5ish.
    """
```

File: astroid/brain/brain_qt.py [hunk 9]
```diff
@@ -71,6 +71,17 @@ def emit(self, signal): pass
...
+
```
Comment (DanielNoord) [inline astroid/brain/brain_qt.py]:
```suggestion
```

**Review by DanielNoord — CHANGES_REQUESTED** [review]
Sorry for pushing to your branch directly. I saw we had a very similar function so I thought it made sense just to merge the two and keep the detection of `Signal` in one place.

Would you be able to write a test for this? The rest of the `qt` brain is also covered, so there should be some pre-existing tests.

**Comment by DanielNoord** [comment]
> > Would you be able to write a test for this? The rest of the qt brain is also covered, so there should be some pre-existing tests.
> 
> Yes, it is covered, but only with PyQt6 as the Qt library, not PySide2 or 6. Would you propose to add PySide6 as an additional test requirement?

That's perhaps a bit much. Should we add coverage pragma's then?

**Comment by adam-grant-hendry** [comment]
I'll push my test up for your review.

File: requirements_test_brain.txt [hunk 10]
```diff
@@ -3,7 +3,10 @@ types-attrs
 nose
 numpy>=1.17.0; python_version<"3.11"
 python-dateutil
+PyQt5>=5.15.2
```
Comment (Pierre-Sassoulas) [inline requirements_test_brain.txt:6]: I think we can all-in with the brain test requirements and add major libraries. We have very expansive primer for pylint, brain test are comparatively very cheap for astroid. We probably need to define what constitute a major library and what should be maintained in an astroid-plugin though.

**Review by Pierre-Sassoulas — APPROVED** [review]
LGTM, nice merge request !

**Comment by adam-grant-hendry** [comment]
> I was thinking about test for the qt brain.

Ah yes, I see. Yes, I've added a regression test that should works for `PyQt5`, `PySide2`, `PyQt6`, and `PySide6`. It requires adding the following to `requirements_test_brain.txt` though, just FYI, so unfortunately to your other question no: additional dependencies must be installed:

### requirements_test_brain.txt

```python
...
PyQt5>=5.15.2
PySide2>=5.15.2
...
PySide6
```

**Comment by Pierre-Sassoulas** [comment]
> How would you like to proceed?

Let's use version specifiers so pip does not fail. The tests will be skipped on 3.11 because the dependencies won't be installed. We can remove the specifier when it's compatible with 3.11 later.

**Review by adam-grant-hendry — COMMENTED** [review]
LGTM. Will let CI re-run and go forward from there.

**Comment by adam-grant-hendry** [comment]
> > How would you like to proceed?
> 
> Let's use version specifiers so pip does not fail. The tests will be skipped on 3.11 because the dependencies won't be installed. We can remove the specifier when it's compatible with 3.11 later.

@Pierre-Sassoulas Perhaps unrelated, but I'm on Windows 10 and am also noticing when running `tox` that python 3.7-3.10 all fail 3 tests:

```
tests\unittest_decorators.py::TestDeprecationDecorators.test_deprecated_default_argument_values_one_arg
tests\unittest_decorators.py::TestDeprecationDecorators.test_deprecated_default_argument_values_two_args
tests\unittest_decorators.py::test_deprecation_of_doc_attribute
```

Here's the output:

[astroid_PR1654_errors.txt](https://github.com/PyCQA/astroid/files/8980645/astroid_PR1654_errors.txt)

File: tests/unittest_brain_qt.py [hunk 11]
```diff
@@ -93,7 +93,7 @@ def test_slot_disconnect_no_args() -> None:
     reason="These tests require the PyQt5, PySide2, PyQt6, and PySide6 libraries.",
 )
 @pytest.mark.skipif(
-    PY311_PLUS, reason="PySide2>=5.15.2 not yet available for Python 3.11"
+    PY311_PLUS, reason="PySide2>=5.15.2 and PySide6 not yet available for Python 3.11"
 )
```
Comment (Pierre-Sassoulas) [inline tests/unittest_brain_qt.py:97]: I think this is unnecessary as we already have line 88 a,d 90 to handle that.

Comment (adam-grant-hendry) [reply tests/unittest_brain_qt.py]: Ah, I see. Instead I should have simply put the restriction in `requirements_test_brain.txt`:

```
PyQt5>=5.15.2
PySide2>=5.15.2; python_version<"3.11"
PyQt6
PySide6; python_version<"3.11"
```

I'll remove this skip line.

**Comment by adam-grant-hendry** [comment]
> > Sorry for pushing to your branch directly. I saw we had a very similar function so I thought it made sense just to merge the two and keep the detection of `Signal` in one place.
> > Would you be able to write a test for this? The rest of the `qt` brain is also covered, so there should be some pre-existing tests.
> 
> Quite alright, no worries. I'll fix the docstring too.

I also hope I haven't inadvertently overwritten any of your suggestions. I don't believe I have, but let me know if something got missed. I had to rebase a couple times to pull in your latest changes.

**Comment by DanielNoord** [comment]
> I also hope I haven't inadvertently overwritten any of your suggestions. I don't believe I have, but let me know if something got missed. I had to rebase a couple times to pull in your latest changes.

Sadly you did 😄 But no worries, I'll repush. You're also more than welcome to use `git pull` and don't rebase. For small, related changes such as this we tend to squash merge anyway!

**Comment by adam-grant-hendry** [comment]
@Pierre-Sassoulas Python 3.7 to 3.9 continue to segfault. I notice `RuntimeWarning` messages in 3.10, which I've encountered before:

`shiboken6`, which creates the python bindings for `Qt` C++ source, imports from a `zip` file into the top-level directory at runtime. These files are deleted after running, but `coverage` attempts to look at their source after they're gone, causing
warnings to appear. Namely, it looks for these modules/files:

   project_dir/pysrcript
   project_dir/shibokensupport
   project_dir/signature_bootstrap.py

To avoid this error, `source` can be specified to the package subdirectory (i.e. `astroid/`). However, this can also be avoided by explicitly omitting these folders in the `omit` section for coverage.

See https://github.com/nedbat/coveragepy/issues/1392

**Comment by adam-grant-hendry** [comment]
> > I also hope I haven't inadvertently overwritten any of your suggestions. I don't believe I have, but let me know if something got missed. I had to rebase a couple times to pull in your latest changes.
> 
> Sadly you did 😄 But no worries, I'll repush. You're also more than welcome to use `git pull` and don't rebase. For small, related changes such as this we tend to squash merge anyway!

I'm so sorry about that!

**Comment by adam-grant-hendry** [comment]
> @Pierre-Sassoulas Python 3.7 to 3.9 continue to segfault. I notice `RuntimeWarning` messages in 3.10, which I've encountered before:
> 
> `shiboken6`, which creates the python bindings for `Qt` C++ source, imports from a `zip` file into the top-level directory at runtime. These files are deleted after running, but `coverage` attempts to look at their source after they're gone, causing warnings to appear. Namely, it looks for these modules/files:
> 
> project_dir/pysrcript project_dir/shibokensupport project_dir/signature_bootstrap.py
> 
> To avoid this error, `source` can be specified to the package subdirectory (i.e. `astroid/`). However, this can also be avoided by explicitly omitting these folders in the `omit` section for coverage.
> 
> See [nedbat/coveragepy#1392](https://github.com/nedbat/coveragepy/issues/1392)

@DanielNoord Perhaps you'd like to review this as well. I'm not sure what's causing the segfault errors I'm seeing.

**Comment by DanielNoord** [comment]
I haven't followed the discussion, but these test (that I originally asked for 😓) seem to be causing a lot of issues.

I would be fine with your confirmation that the latest version of this PR fixes the issue and adding some `pragma: no cover` comments on the code that is not covered. This seems like an awful lot of trouble for a simple `str() in list()` check.

**Review by adam-grant-hendry — COMMENTED** [review]
Oh this is much cleaner, thank you! I like it.

File: astroid/brain/brain_qt.py [hunk 12]
```diff
@@ -11,10 +11,14 @@
 
 
 def _looks_like_signal(node, signal_name="pyqtSignal"):
```
Comment (DanielNoord) [inline astroid/brain/brain_qt.py:13]:
```suggestion
def _looks_like_signal(node: nodes.FunctionDef, signal_name: str = "pyqtSignal") -> bool:
```

Comment (adam-grant-hendry) [reply astroid/brain/brain_qt.py]: LGTM!

**Comment by adam-grant-hendry** [comment]
> I haven't followed the discussion, but these test (that I originally asked for 😓) seem to be causing a lot of issues.
> 
> I would be fine with your confirmation that the latest version of this PR fixes the issue and adding some `pragma: no cover` comments on the code that is not covered. This seems like an awful lot of trouble for a simple `str() in list()` check.

Oh nice! Let me try one more thing though. I noticed I accidentally altered the original `class TestBrainQt` to skip if `HAS_PYSIDE6` when it should be `HAS_PYQT6`.

I'd like to run the CI one more time and if that doesn't fix it, I'll remove the test, changes to `requirements_test_brain.txt`, and add the `# pragma: no cover`s.

**Comment by adam-grant-hendry** [comment]
> > I haven't followed the discussion, but these test (that I originally asked for 😓) seem to be causing a lot of issues.
> > I would be fine with your confirmation that the latest version of this PR fixes the issue and adding some `pragma: no cover` comments on the code that is not covered. This seems like an awful lot of trouble for a simple `str() in list()` check.
> 
> Oh nice! Let me try one more thing though. I noticed I accidentally altered the original `class TestBrainQt` to skip if `HAS_PYSIDE6` when it should be `HAS_PYQT6`.
> 
> I'd like to run the CI one more time and if that doesn't fix it, I'll remove the test, changes to `requirements_test_brain.txt`, and add the `# pragma: no cover`s.

Okay, unfortunately that didn't work, so I reverted the test and `requirements_test_brain.txt` and added the `# pragma: no cover`s. Hopefully CI should pass now.

**Comment by adam-grant-hendry** [comment]
@DanielNoord Would you be able to see if CI will now honor the `# pragma: no cover`s?

**Comment by adam-grant-hendry** [comment]
> Sorry for pushing to your branch directly. I saw we had a very similar function so I thought it made sense just to merge the two and keep the detection of `Signal` in one place.
> 
> Would you be able to write a test for this? The rest of the `qt` brain is also covered, so there should be some pre-existing tests.

I believe this was completed, but is still being shown as not resolved (FYI).

**Review by jacobtylerwalls — APPROVED** [review]

**Review by Pierre-Sassoulas — APPROVED** [review]
Sorry for the tests that were so hard to do. Thank you for all the work and taking remarks into account @adam-grant-hendry.

**Review by DanielNoord — APPROVED** [review]

**Pull request merged**